src/fides/api/ops/api/v1/endpoints/user_permission_endpoints.py

```diff
@@ -79,13 +85,21 @@ def update_user_permissions(
     user_id: str,
     permissions: UserPermissionsEdit,
 ) -> FidesUserPermissions:
+    """Update either a user's role(s) and/or scope(s).
+
+    Typically we'll assign roles to a user and they'll inherit the associated scopes,
+    but we're still supporting assigning scopes directly as well.
+    """
     user = validate_user_id(db, user_id)
     logger.info("Updated FidesUserPermission record")
+
     if user.client:
-        user.client.update(db=db, data={"scopes": permissions.scopes})
+        user.client.update(
+            db=db, data={"scopes": permissions.scopes, "roles": permissions.roles}
+        )
     return user.permissions.update(  # type: ignore[attr-defined]
         db=db,
-        data={"id": user.permissions.id, "user_id": user_id, **permissions.dict()},  # type: ignore[attr-defined]
+        data={"id": user.permissions.id, "user_id": user_id, "scopes": permissions.scopes, "roles": permissions.roles},  # type: ignore[attr-defined]
```

Comment:
This is the endpoint that allows you to update a user's permissions. On `main` you can add scopes to users, this additionally allows you to assign roles (which are associated with scopes).   We are still going to support being assigned both scopes and roles.  (For one, you can currently only assign scopes directly in the UI, but two, Sean said we should continue to allow scopes to be assigned for API-driven workflows).

Reply:
What if a user's role and scopes conflict with each other? what takes precedence and how will end users know what this precedence is?

Either way I can see this being confusing to users/administrators.  Lets say you and I both are in the `VIEWER` group and you have an added scope to also approve. You ask me to approve for you, but I get an error. It totally makes sense why I get an error, but to the end user they are going to think they are in the same group so they can do the same thing. We will need to have a way to make it clear to the end users what happened in situations like this.

Reply:
Thanks for mentioning this, I've been mulling over this too!  I've also been concerned that since the "scopes" UI is going away, there's going to be no easy visibility into who has these extra scopes.   From our refinement discussion on this yesterday, however, the majority opinion seemed to be that this was an edge case and we could address if it became an issue.


- We still need to support scopes generally for people just using the API, for example someone developing an app against a fides instance or someone that wants to create a very limited token to do one specific job.
- Regardless, we need to support both roles and scopes on users as a transitional measure as there is no "roles" UI yet. 
- I suggested a future ticket to put in the backlog where we **deprecated assigning scopes directly to users, but still allowed scopes to be assigned directly to clients.**  (So the user permissions table would lose the scopes column, and clients created/updated from user permissions would only have roles persisted to the client and then the token.  Folks using the API-only workflow could still create clients directly and assign scopes directly to those clients).  Would something like that help here @sanders41?

Reply:
> What if a user's role and scopes conflict with each other? what takes precedence and how will end users know what this precedence is?

Regarding "precedence", there aren't really conflicts here.  Permission checking just looks to see "do you have this scope through some avenue?" You might have the scope because you have a specific role, or you might have the scope because you have it directly.  You might be granted the scope from both places and that's fine.

What you can see is _really the sum of all of your scopes from all your sources._  

In practice, with the roles-based UI going in, most users will have a single role assigned to them, and inherit the associated list of scopes.

Reply:
Yes, I think this would solve the issue.